cmd/metal-api/internal/service/tenant-service.go

```diff
@@ -91,3 +133,132 @@ func (r *tenantResource) listTenants(request *restful.Request, response *restful
+		return
+	}
+
+	res, err := r.mdc.Tenant().Find(context.Background(), mapper.ToMdmV1TenantFindRequest(&requestPayload))
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	var ps []*v1.Tenant
+	for i := range res.Tenants {
+		v1p := mapper.ToV1Tenant(res.Tenants[i])
+		ps = append(ps, v1p)
+	}
+
+	r.send(request, response, http.StatusOK, ps)
+}
+
+func (r *tenantResource) createTenant(request *restful.Request, response *restful.Response) {
+	var pcr v1.TenantCreateRequest
+	err := request.ReadEntity(&pcr)
+	if err != nil {
+		r.sendError(request, response, httperrors.BadRequest(err))
+		return
+	}
+
+	tenant := mapper.ToMdmV1Tenant(&pcr.Tenant)
+
+	mdmv1pcr := &mdmv1.TenantCreateRequest{
+		Tenant: tenant,
+	}
+
+	p, err := r.mdc.Tenant().Create(context.Background(), mdmv1pcr)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	v1p := mapper.ToV1Tenant(p.Tenant)
+	pcres := &v1.TenantResponse{
+		Tenant: *v1p,
+	}
+
+	r.send(request, response, http.StatusCreated, pcres)
+}
+
+func (r *tenantResource) deleteTenant(request *restful.Request, response *restful.Response) {
+	id := request.PathParameter("id")
+
+	pgr := &mdmv1.TenantGetRequest{
+		Id: id,
+	}
+	p, err := r.mdc.Tenant().Get(context.Background(), pgr)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	plr, err := r.mdc.Project().Find(context.Background(), &mdmv1.ProjectFindRequest{TenantId: wrapperspb.String(id)})
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+	if len(plr.Projects) > 0 {
+		r.sendError(request, response, httperrors.BadRequest(errors.New("there are still projects allocated by this tenant")))
+		return
+	}
+
+	pdr := &mdmv1.TenantDeleteRequest{
+		Id: p.Tenant.Meta.Id,
+	}
+	pdresponse, err := r.mdc.Tenant().Delete(context.Background(), pdr)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	v1p := mapper.ToV1Tenant(pdresponse.Tenant)
+	pcres := &v1.TenantResponse{
+		Tenant: *v1p,
+	}
+
+	r.send(request, response, http.StatusOK, pcres)
+}
+
+func (r *tenantResource) updateTenant(request *restful.Request, response *restful.Response) {
+	var requestPayload v1.TenantUpdateRequest
+	err := request.ReadEntity(&requestPayload)
+	if err != nil {
+		r.sendError(request, response, httperrors.BadRequest(err))
+		return
+	}
+
+	if requestPayload.Tenant.Meta == nil {
+		r.sendError(request, response, httperrors.BadRequest(errors.New("tenant and tenant.meta must be specified")))
+		return
+	}
+
+	existingTenant, err := r.mdc.Tenant().Get(context.Background(), &mdmv1.TenantGetRequest{Id: requestPayload.Tenant.Meta.Id})
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
+
+	// new data
+	tenantUpdateData := mapper.ToMdmV1Tenant(&requestPayload.Tenant)
+	// created date is not updateable
+	tenantUpdateData.Meta.CreatedTime = existingTenant.Tenant.Meta.CreatedTime
```

Comment:
From my perspective either the server side should deal with the case when the `CreatedTime` field is empty and apply this default. Or the server requires to always send all fields on updates but then metalctl needs to take care of this.